### ISSUE TITLE
Files with ? on an NFS share are truncated when inserted into files table in MySQL

### ISSUE THREAD

**Issue by kaelspencer** [xbmc/xbmc#16383]
### Describe the bug
When you use a combination of NFS shares, MySQL server for library information, and filenames with a question mark (?) in them, the entry into the *files* table is truncated at the question mark. This results in an entry in the library that is not playable, but will be detected each time a scan is performed if you remove the "not found" file.

Major assumption number one: ? is a valid character in a file on non-Windows systems. Number two: ? is valid in an NFS path (I can browse to and play these files outside of Kodi).

## Expected Behavior
I expect the full filename, including question mark, to be inserted into the *files* table

## Actual Behavior
A truncated path is inserted into the *files* table.

## Possible Fix
Based on the logs below, I think this is what's happening.

1. Full path is provided to `CVideoDatabase::AddFile`, param strFileNameAndPath is "nfs://10.10.0.11/export/mediaro/Video/TV Shows/Foo/Season 01/Foo - S01E02 - Special? Character.mkv".
2. `CVideoDatabase::SplitPath` is called.
3. `URIUtils::Split` is called. When `IsURL` is called, it is returning true because the path contains "://". While walking the path, it hits ? and breaks. This lops off the end of the filename in this case.

This is similar to what [trac 17627](https://trac.kodi.tv/ticket/17627) was, and then that was fixed with PR #12962. However, the NFS path is getting caught in the IsURL check.

I don't know all the other use cases, but I don't think `IsURL` should return true for nfs://.

### To Reproduce
1. Have sources on an NFS share.
2. Have a filename with a ? in it.
3. Use MySQL to you consume CVideoDatabase (I think this is important, but I've never not used MySQL).

### Debuglog
The debuglog can be found here: https://paste.kodi.tv/evogelubev.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem: none.

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

Some relevant log snippets.

```
2019-07-18 19:39:00.191 T:140429511034624   DEBUG: VideoInfoScanner: Found episode match nfs://10.10.0.11/export/mediaro/Video/TV Shows/Foo/Season 01/Foo - S01E01 - Basic filename.mkv (s1e1) [s([0-9]+)[ ._x-]*e([0-9]+(?:(?:[a-i]|\.[1-9])(?![0-9]))?)([^\\/]*)$]
2019-07-18 19:39:00.191 T:140429511034624   DEBUG: VideoInfoScanner: Found episode match nfs://10.10.0.11/export/mediaro/Video/TV Shows/Foo/Season 01/Foo - S01E02 - Special? Character.mkv (s1e2) [s([0-9]+)[ ._x-]*e([0-9]+(?:(?:[a-i]|\.[1-9])(?![0-9]))?)([^\\/]*)$]

2019-07-18 19:39:00.751 T:140429511034624   DEBUG: Mysql execute: insert into path (idPath, strPath, idParentPath) values (NULL, 'nfs://10.10.0.11/export/mediaro/Video/TV Shows/Foo/Season 01/', 4931)

2019-07-18 19:39:00.823 T:140429511034624   DEBUG: VideoInfoScanner: Adding new item to tvshows:nfs://10.10.0.11/export/mediaro/Video/TV Shows/Foo/Season 01/Foo - S01E01 - Basic filename.mkv
2019-07-18 19:39:00.824 T:140429511034624   DEBUG: Mysql execute: insert into files (idFile, idPath, strFileName) values(NULL, 4932, 'Foo - S01E01 - Basic filename.mkv')


2019-07-18 19:39:01.163 T:140429511034624   DEBUG: VideoInfoScanner: Adding new item to tvshows:nfs://10.10.0.11/export/mediaro/Video/TV Shows/Foo/Season 01/Foo - S01E02 - Special? Character.mkv
2019-07-18 19:39:01.164 T:140429511034624   DEBUG: Mysql execute: insert into files (idFile, idPath, strFileName) values(NULL, 4932, 'Foo - S01E02 - Special')
```

The important log lines are the last two. First, you see that VideoInfoScanner logs out the full path. Then you see that the truncated path was inserted into *files*.

What's actually in the database, you ask?

```
mysql> select * from files where idPath=4932;
+--------+--------+-----------------------------------+-----------+------------+---------------------+
| idFile | idPath | strFilename                       | playCount | lastPlayed | dateAdded           |
+--------+--------+-----------------------------------+-----------+------------+---------------------+
|  44141 |   4932 | Foo - S01E01 - Basic filename.mkv |      NULL | NULL       | 2019-07-18 19:36:47 |
|  44142 |   4932 | Foo - S01E02 - Special            |      NULL | NULL       | 2019-07-18 19:37:01 |
+--------+--------+-----------------------------------+-----------+------------+---------------------+
```

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Ubuntu 19.04
 - Kodi version: 18.3

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by KarellenX** [comment]
Another issue along similar lines... 
https://github.com/xbmc/xbmc/issues/16114

**Comment by wsnipex** [comment]
? is a reserved character on all windows and older macos file systems, so fixing this in a portable way is not really possible

**Comment by kaelspencer** [comment]
I don't think the issue is about cross platform support. Instead, I think the issue is that NFS is presenting a valid file path to the system, but CVideoDatabase is _incorrectly_ stripping characters from that path and inserting a broken path into the database.

Specifically, this code below is incorrect for NFS. I'd argue that it's not even necessary for SMB files because they won't have a ? in them anyway.

https://github.com/xbmc/xbmc/blob/55d75d979a161bbfcfc8b31bb5272f4d2934f42f/xbmc/utils/URIUtils.cpp#L225-L236

**Comment by Uatschitchun** [comment]
I would like to add that I'm stumbling upon the same issue!

To add, it's an issue only appearing when using libnfs! If using kernel nfs (nfs share mounted to the file system) all files with special characters are recognized fine and play well!

Even if using libnfs, those files play well, when opening them in file mode. 

This is an unnecessary restriction in using libnfs.

**Comment by Uatschitchun** [comment]
It's not only a matter of "component: Video", but is valid for music, too!
https://forum.kodi.tv/showthread.php?tid=349284&pid=2902280#pid2902280

**Comment by Uatschitchun** [comment]
> ? is a reserved character on all windows and older macos file systems, so fixing this in a portable way is not really possible

Isn't it the job of the underlying file system to take care of allowed characters? It's about reading files, not writing them..?

**Comment by Uatschitchun** [comment]
Btw, a path like:
/media/music/unsorted?/working?.mp3
with /media/music being a system mounted nfs-share, 
whereas a path like:
nfs://192.168.1.1/media/music/unsorted?/working?.mp3
only works in file mode, but gets truncated when added to library m

**Comment by rugalash** [comment]
Can confirm what the above says. Files and NFS are working just fine, Kodi "add to library" is stripping the ? and everything after from what it looks like, for no real reason either. None of these names are being added to the client device at all other than in the database which handles it fine.

**Comment by kaelspencer** [comment]
As far as I can tell, the fix is quite easy. I think `IsURL` should be returning false for NFS and SMB shares. I haven't had the time to get setup to build and test Kodi changes.

**Comment by bernimoses** [comment]
Is there any other schema (ftp, udp, ...) where `?` has a special meaning like with http(s)? If not maybe the [`IsURL`](https://github.com/xbmc/xbmc/blob/55d75d979a161bbfcfc8b31bb5272f4d2934f42f/xbmc/utils/URIUtils.cpp#L225-L236) should be switched to `IsHTTP` to avoid other problems as well.

**Comment by yol** [comment]
Is this reproducible without MySQL? If not, the explanations given might not be valid.

**Comment by yol** [comment]
> As far as I can tell, the fix is quite easy. I think IsURL should be returning false for NFS and SMB shares. I haven't had the time to get setup to build and test Kodi changes.

I think it would be better to only have URLs and local paths. With this separation, you would create a third category of paths that look like URLs but kind of aren't.

**Comment by DaveTBlake** [comment]
>I expect the full filename, including question mark, to be inserted into the files table

**Comment by DaveTBlake** [comment]
>I expect the full filename, including question mark, to be inserted into the files table

There are a number of charaters including ? (question mark)  that are not acceptable as filenames across all platforms. For example Windows has the following reserved characters:
        < (less than)
        > (greater than)
        : (colon)
        " (double quote)
        / (forward slash)
        \ (backslash)
        | (vertical bar or pipe)
        ? (question mark)
        * (asterisk)

It seems likely that Kodi was designed with a similar restriction on filenames

**Comment by yol** [comment]
> There are a number of charaters including ? (question mark) that are not acceptable as filenames across all platforms.

That is not true. POSIX (including Linux and therefore Android) allows everything except forward slash and NUL. OS X is basically the same (reportedly, HFS+ file names can even contain NUL). It is only Windows that is this restrictive.

The issue here is the URL encoding, in which `?` has special meaning. This problem will never occur on Windows since it disallows the same character in file names. I think this is by chance and not by design.

**Comment by rugalash** [comment]
@DaveTBlake Please do not post incorrect information and make stuff up that you do not know. Here or anywhere else. Just do not speak if you don't know, that is better than speaking wrong.
![image](https://user-images.githubusercontent.com/56284636/84655743-1adb4200-aed7-11ea-9fe5-a0d537984479.png)

**Comment by DaveTBlake** [comment]
@rugalash are refuting that Windows has the reserved char limitation on filenames I listed? Hence not all platforms accecpt all chararcters in filenames. **Both are facts**.

My hypothesis that Kodi was designed with a similar restriction on filenames may well be wrong, it could easily infact be by chance, but I don't think that deserves your kind of reponse. I was in discussion with team members about how we could best fix this problem, it is obviously best I don't waste my time on this any further.

**Comment by tadly** [comment]
After some more discussions internally it has been pretty much agreed upon that changing `URIUtils::Split` is **not** the way to go.

Instead, `filesystem/NFSDirectory.cpp` and possibly `filesystem/NFSFile.cpp` would need a rework as they handle url related things as plain strings, (concatenating them using `+` etc.) instead of using proper CURL instances.
As such any form of character encoding is missing, causing the outlined issue.

**Comment by Copyrites** [comment]
What's the priority on this bug?  Its the only thing holding me back from upgrading.

**Comment by Uatschitchun** [comment]
Nice to see some progress on solving this issue

**Comment by tadly** [comment]
So, because I really wanted to continue a show I thought I give sftp a shot.

It scans into the DB when you add a new source or do a "library update" (unlike NFS in regards to the later IIRC) but does also **not** play.
A subsequent "Clean Library" will remove it just like with NFS.

What I'm saying is, there's more to be done then refactoring the NFS stuff

**Comment by kaelspencer** [comment]
@tadly I think that's the bug, right? The scanner is detecting the new file but puts the incorrect path in the db. Because the path is incorrect, you can't play it. Because the path doesn't actually exist, "Clean Library" removes that entry.

I'm still of the view that if the sources (i.e., NFS server) says the path is _foo_, then Kodi should put _foo_ in the db

**Comment by tadly** [comment]
@kaelspencer yes, this is the bug. The problem however is (as mentioned before) the way the url is handled/passed around inside the code, not that one part in `URIUtils.cpp`

Because NFS (and probably other parts in core) handle urls as **string** rather than **curl objects** we end up with stuff like:
`str: nfs:///path/to/movies/my?movie.mkv` (invalid)

rather then:
`str: nfs://path/to/movies/my%3FBmovie.mkv` (valid)

**Comment by yol** [comment]
I'll be brutally honest - since there is an easy workaround (not have `?` in the filename which would break on many FS anyway), I think no one currently has this high on the priority list :-(

**Comment by tadly** [comment]
I agree that changing the filename is easier but isn't it only windows ntfs which doesn't support `?`? (and maybe other filesystems that no one uses anyways)
I'd even argue that the vast majority of Kodi users have their media stored on a filesystem which does support `?` (alongside other characters).

Also, given how the whole url handling issue seemingly isn't limited to NFS, I'd consider this to be quite a major issue (objectively speaking).

I understand that this is probably a huge undertaking and given the possible workaround I also understand the lack of interest in working on this.
Just wanted to add my opinion :)

**Comment by yol** [comment]
> I agree that changing the filename is easier but isn't it only windows ntfs which doesn't support ?? (and maybe other filesystems that no one uses anyways)
> I'd even argue that the vast majority of Kodi users have their media stored on a filesystem which does support ? (alongside other characters).

It'll break as soon as you want to copy the file to an external (USB) drive though, which to 99% use NTFS or exFAT (which has similar limitations) :shrug:

> Also, given how the whole url handling issue seemingly isn't limited to NFS, I'd consider this to be quite a major issue (objectively speaking).

The only other major remote protocol is SMB, which usually doesn't allow `?` as well already on the server side. (Windows server will not allow it for obvious reasons, and samba in default config also does not allow it)

Of course, this is a valid bug. But from a user POV, I'd still argue that, in the long run, you're better off avoiding `?` and similar in file names.

**Comment by Copyrites** [comment]
> The only other major remote protocol is SMB, which usually doesn't allow ? as well already on the server side. 

My impression is that the majority of consumer-grade mediaservers run on a unix rather than a windows so for the most common kodi use case, most SMB servers will allow the ? in the filename.

>  But from a user POV, I'd still argue that, in the long run, you're better off avoiding ? and similar in file names.

Two points to that:

1. It makes automation difficult - generating filenames from music tags and tv/movie databases.  Speaking from experience there are a lot more titles with question marks in the name than seems intuitive.  So, at best, it becomes an obscure setting the user has to become knowledgeable.

2. Discoverability is horrible.  It just silently fails without any informative messages.  That wastes tons of man hours across all of the people trying to figure it out.

**Comment by yol** [comment]
> My impression is that the majority of consumer-grade mediaservers run on a unix rather than a windows

Definitely

> so for the most common kodi use case, most SMB servers will allow the ? in the filename.

That depends upon the configuration (no idea how manufacturers set up their samba)

> It makes automation difficult - generating filenames from music tags and tv/movie databases

FWIW, most software that deals with this (opensource or otherwise) does not generate file names with `?` even on UNIX or at least has an option to not do so

> Discoverability is horrible. It just silently fails without any informative messages.

Very valid point.

Anyway, fighting about the severity will not get us much closer to a fix, so I'll stop now.

**Comment by Uatschitchun** [comment]
As a reminder, kodi's handling of filenames with "?" works very well when using kernel NFS instead of libNFS. 
It's not about one character only, but about keeping conformity with standards. So, if there is support within FS and within NFS kodi should support it, too!

**Comment by wsnipex** [comment]
PRs welcome

**Comment by pprkut** [comment]
@DaveTBlake mentioned in #18311 that the original suggested fix (https://github.com/xbmc/xbmc/issues/16383#issuecomment-644137087) doesn't look like the right option after all.

Instead, we're back at looking at fixing the issue in a generic, non-NFS specific way. @DaveTBlake mentioned an option in https://github.com/xbmc/xbmc/pull/18311#issuecomment-903235770.

Based on what I understand we could use `CURL::GetFileNameWithoutPath()`, `CURL::GetWithoutFilename()` and potentially `CURL::GetWithoutOptions()` instead of `URIUtils::Split()`. I only have limited understanding of the code dependencies here, but if we only fix this for VideoDatabase/MusicDatabase, wouldn't it be still a problem in other places where `URIUtils::Split` is used? Wouldn't it be better to adjust the options removal in `URIUtils::Split` after all?

**Comment by pprkut** [comment]
Well, having looked at the code, turns out it's not that simple after all. `CURL` isn't aware of either the `file` or `nfs` protocol right now. That is easy enough to fix, but the option parsing will consider everything after the first `?` as an option. So a path like `nfs:///path/to?/movie?.avi?showinfo=true` will be truncated to `nfs:///path/to`, which if I understand the current code correctly would be even worse than what we have now (which should be `nfs:///path/to?/movie`.

The paths parsed by the `CURL` class would need to be properly URL encoded for the detection to work properly.

**Comment by DaveTBlake** [comment]
Sorry @pprkut for looking at this a year later and requestioning the route to take, but seeing the encoded data in the db made some of the other issues that will cause more obvious to me. 

In the libraries as data for querying etc. path and filename want to be strings that a human can read, how a user would expect the folder(s) and file name of the media to be displayed, that is what we currently have. These stored path and filename values are also recombined to make a Uniform Resource Identifier (URI) used to identify the media resource to play. Kodi now handles very diverse media locations, not only local files or over LAN but streamed from plugins, served over WAN etc. where protocol options  e.g. `http://localhost/a|b=c` (but not query options e.g. `http://localhost/a?b=c` AFAIK)  are needed too. Protocol options get stored with the path.

The growing complexity of the URI Kodi handles, with possible special charaters in unexpected places, implied storing encoded URLs rather than reconstruct them could be a good idea.  However we still also need path and filename for items that have them in a decoded format as library data. That really is a bigger question and far reaching change than handling `?` in filenames on NFS.

It makes me think that if handling `?` in filenames (there is no problem with `?` in the path) in all sources (not just NFS) is something we want Kodi to support then this is best done somewhere in the "split" processing.  My bottom line is that the databases need to keep their decoded strings for path and filename.

**Comment by DaveTBlake** [comment]
> but the option parsing will consider everything after the first ? as an option.

Oddly I didn't see that in the brief testing I did @pprkut 
With `nfs:///path/to?/movie?.avi` `CURL::GetProtocolOptions()` does not find any options and `CURL::GetFileNameWithoutPath()` returns `movie?.avi`

**Comment by pprkut** [comment]
That is correct, because there's no separator characters defined for the NFS protocol. If you test with `nfs:///path/to?/movie?.avi?showinfo=true` then `CURL::GetFileNameWithoutPath()` will return `movie?.avi?showinfo=true`. I suppose that's not what we want either.

**Comment by DaveTBlake** [comment]
Ah yes I see. Well if `?` is going to be valid in filenames from NFS then query protocols are not going to be allowed for it, or NFS must have some other query protocol separator. What would add `?showinfo=true` or any other query option to a URI on NFS anyway? A plugin? Maybe nothing.

**Comment by pprkut** [comment]
If we we don't care about query params with `?` for NFS, then the fix should indeed be fairly straight forward. However, officially the syntax would be support, according to https://github.com/sahlberg/libnfs:

```
nfs://<server|ipv4|ipv6>[:<port>]/path[?arg=val[&arg=val]*]
```

**Comment by pprkut** [comment]
Alright, researched further, and landed [here](https://datatracker.ietf.org/doc/html/rfc2224) and [here](https://www.openldap.org/lists/openldap-technical/201208/msg00062.html). According to both those links, the paths in `nfs://` URLs must be US-ASCII, so we would also need to encode other unicode characters (like CJK for example) to be conform), which, I think, would lead us back to the already open pull request :-/

**Comment by DaveTBlake** [comment]
> However, officially the syntax would be supported, according to https://github.com/sahlberg/libnfs: ` nfs://<server|ipv4|ipv6>[:<port>]/path[?arg=val[&arg=val]*]`

Interesting, but that syntax can not work unless `?` is a reserved separator not allowed in path. 
RFC2224 says that "An NFS URL does not include user or password components, nor does it recognize "?" query or "#" fragment  components". Hum....

**Comment by DaveTBlake** [comment]
>  According to both those links, the paths in `nfs://` URLs must be US-ASCII, so we would also need to encode other unicode characters (like CJK for example) to be conform), which, I think, would lead us back to the already open pull request :-/

Does Kodi need to be fussy that the URL it was given as a path and subsequently uses to start playback etc. is US-ASCII? I don't see why it needs to worry. The databases certainly don't care. The question is still `?` as a separator. RFC2224 says not, but libnfs does (maybe?). Those same old docs also say in the BNF section for NFS schema that 
`reserved      = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+"`

**Comment by pprkut** [comment]
libnfs does support them:

```
Libnfs uses RFC2224 style URLs extended with some minor libnfs extensions.
[..]
Arguments supported by libnfs are :
 tcp-syncnt=<int>  : Number of SYNs to send during the session establish
                     before failing setting up the tcp connection to the
                     server.
 uid=<int>         : UID value to use when talking to the server.
                     default it 65534 on Windows and getuid() on unixen.
 gid=<int>         : GID value to use when talking to the server.
                     default it 65534 on Windows and getgid() on unixen.
 readahead=<int>   : Enable readahead for files and set the maximum amount
                     of readahead to <int> bytes.
 auto-traverse-mounts=<0|1>
                   : Should libnfs try to traverse across nested mounts
                     automatically or not. Default is 1 == enabled.
 dircache=<0|1>    : Disable/enable directory caching. Enabled by default.
 autoreconnect=<-1|0|>=1>
                   : Control the auto-reconnect behaviour to the NFS session.
                    -1 : Try to reconnect forever on session failures.
                         Just like normal NFS clients do.
                     0 : Disable auto-reconnect completely and immediately
                         return a failure to the application.
                   >=1 : Retry to connect back to the server this many
                         times before failing and returing an error back
                         to the application.
 if=<interface>    : Interface name (e.g., eth1) to bind; requires `root`
 version=<3|4>     : NFS Version. Default is 3.
 nfsport=<port>    : Use this port for NFS instead of using the portmapper.
 mountport=<port>  : Use this port for the MOUNT protocol instead of
                     using portmapper. This argument is ignored for NFSv4
                     as it does not use the MOUNT protocol.```

**Comment by the-black-eagle** [comment]
No, it doesn't.  Read it properly.  You cannot have `?` in the file name/path if you are using `?` as a delimiter for an option, as in
`nfs://path/to/directory/?version=4

Although `?` is perfectly valid in a local path on Linux, it is not valid in an NFS share path, according to libnfs' documentation or indeed RFC2224.

**Comment by pprkut** [comment]
Sorry, I don't see that in the documentation. If you've found a place where it does say that, can you please post a link?

All I see in RFC2224 is that the url-path needs to be US-ASCII, which includes `?`. `?` is listed as reserved in the BNF, just like `/` and for that one it says:

> Within a <directory> or <name> component the character "/" is reserved and must be encoded as described in Section 2.2 of RFC 1738.

I'm assuming the same holds for the other reserved characters. I find nothing about `?` in paths being invalid in an NFS share path.

As for libnfs, I can't find anything either, but considering that URLs with `?` in the path and or filename work fine, based on earlier reports here and my own testing, leads me to believe that there's at least no limitation in NFS itself to not allow `?`, so it's all about a combination with libnfs's flags. But that should be covered by RFC 2224.

The question left standing is whether kodi wants to support the query parameters of libnfs in file URLs. If not, then we can very easily fix this bug without URL encoding anything, which wouldn't be according to the standard, but if it works...

If we should support them, we would need to encode the directories and filenames, and a db migration that goes along with it.

To be clear, I'm totally fine with taking the easy route and not supporting the query params :)
A middle ground would perhaps be to just encode the `?` character, which would leave the door open to more easily support query params in the future.

**Comment by pprkut** [comment]
@DaveTBlake Not sure #20074 is the fix you imagined, but it works for me :)
(with the caveat of not supporting query params for `nfs://` URLs)

**Comment by DaveTBlake** [comment]
See RCF2224  **BNF for NFS URL Scheme**  https://datatracker.ietf.org/doc/html/rfc2224#section-10 
it says:
`reserved      = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+"`

Also while libnfs says `nfs://<server|ipv4|ipv6>[:<port>]/path[?arg=val[&arg=val]*]` I think it relates to how to give options in a URL input to libnfs such as target version e.g.  `nfs://blah/blah/directory/?version=4`

I don't think any of us read RCF2224 properly a year ago, we got into a discussion of what platforms did or didn't have reserved characters in file naming.

**Comment by the-black-eagle** [comment]
Well, first of all, libnfs' docs clearly state `Libnfs uses RFC2224 style URLs extended with some minor libnfs extensions.`  Those extensions being the ones you quoted in your previous post.

RFC2224 states in section 10 (https://datatracker.ietf.org/doc/html/rfc2224#section-10) what is reserved and what is not.
 
      url-path      = [ "/" ] path_segments
      path_segments = segment *( "/" segment )
      segment       = *pchar
      pchar         = unreserved | escaped | ":" | "@" | "&" | "=" | "+"

      reserved      = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+"
      unreserved    = alpha | digit | mark
      mark          = "$" | "-" | "_" | "." | "!" | "~" |
                      "*" | "'" | "(" | ")" | ","

      escaped       = "%" hex hex
      hex           = digit | "A" | "B" | "C" | "D" | "E" | "F" |
                              "a" | "b" | "c" | "d" | "e" | "f"

      alphanum      = alpha | digit
      alpha         = lowalpha | upalpha

      lowalpha = "a" | "b" | "c" | "d" | "e" | "f" | "g" | "h" | "i" |
                 "j" | "k" | "l" | "m" | "n" | "o" | "p" | "q" | "r" |
                 "s" | "t" | "u" | "v" | "w" | "x" | "y" | "z"
      upalpha  = "A" | "B" | "C" | "D" | "E" | "F" | "G" | "H" | "I" |
                 "J" | "K" | "L" | "M" | "N" | "O" | "P" | "Q" | "R" |
                 "S" | "T" | "U" | "V" | "W" | "X" | "Y" | "Z"
      digit    = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" |
                 "8" | "9"

As I said, just because something is supported via a local file system does not necessarily mean it is supported over nfs.  As Kodi uses libnfs for nfs support, and as libnfs appears to **not** support `?` in filenames/directory names (as per RFC2224) then it seems that Kodi is respecting the standards defined there.  I would much prefer we adhere to such standards as not doing so has in the past come back to bite the team.

**Comment by pprkut** [comment]
Perhaps as additional data point, this works:

```
nfs://192.168.1.147/volume1/Videos/Kabarett/Stermann & Grissemann - Wollt Ihr das totale Sieb?!/Stermann & Grissemann - Wollt Ihr das totale Sieb?!.mkv
```
But this doesn't:
```
nfs://192.168.1.147/volume1/Videos/Kabarett/Stermann & Grissemann - Wollt Ihr das totale Sieb?!/Stermann & Grissemann - Wollt Ihr das totale Sieb%3F!.mkv
```

Just a simple test to see whether libnfs would support the escaped `?`.

**Comment by Copyrites** [comment]
> reserved      = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+"


Filenames using any of these except "/" and "?" currently work with libnfs.

So kodi is not respecting the standards in any consistent way.

**Comment by sahlberg** [comment]
Ouch, this is unfortunate.
I suggest leaving the unescaped '?' as is and let it be the separator between the url and the arguments as it is today.
But we should change :
1, kodi to escape at least '?', but possibly also '&' and '=', when they occur in the /path/file part of the nfs url.
2, libnfs to understand and unescape characters of the form '% hex hex'

I.e. leave the unescaped '?' as url/argument separator
and then escape all '?' in path/filenames when used to form a url.

**Comment by sahlberg** [comment]
I have added support for escaped characters in the URL in the master branch of libnfs.
